crates/eips/src/eip7702/auth_list.rs

```diff
@@ -0,0 +1,206 @@
+use core::ops::Deref;
+
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+use alloy_primitives::{keccak256, Address, ChainId, B256};
+use alloy_rlp::{BufMut, Decodable, Encodable, Header, RlpDecodable, RlpEncodable};
+
+/// An unsigned EIP-7702 authorization.
+#[derive(Debug, Clone, RlpEncodable, RlpDecodable, Eq, PartialEq)]
+pub struct Authorization {
+    /// The chain ID of the authorization.
+    pub chain_id: ChainId,
+    /// The address of the authorization.
+    pub address: Address,
+    /// The nonce for the authorization.
+    pub nonce: OptionalNonce,
+}
+
+impl Authorization {
+    /// Get the `chain_id` for the authorization.
+    ///
+    /// # Note
+    ///
+    /// Implementers should check that this matches the current `chain_id` *or* is 0.
+    pub const fn chain_id(&self) -> ChainId {
+        self.chain_id
+    }
+
+    /// Get the `address` for the authorization.
+    pub const fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Get the `nonce` for the authorization.
+    ///
+    /// # Note
+    ///
+    /// If this is `Some`, implementers should check that the nonce of the authority is equal to
+    /// this nonce.
+    pub fn nonce(&self) -> Option<u64> {
+        *self.nonce
+    }
+
+    /// Computes the signature hash used to sign the authorization, or recover the authority from a
+    /// signed authorization list item.
+    ///
+    /// The signature hash is `keccak(MAGIC || rlp([chain_id, [nonce], address]))`
+    #[inline]
+    pub fn signature_hash(&self) -> B256 {
+        use super::constants::MAGIC;
+
+        #[derive(RlpEncodable)]
+        struct Auth {
+            chain_id: ChainId,
+            nonce: OptionalNonce,
+            address: Address,
+        }
+
+        let mut buf = Vec::new();
+        buf.put_u8(MAGIC);
+
+        Auth { chain_id: self.chain_id, nonce: self.nonce, address: self.address }.encode(&mut buf);
+
+        keccak256(buf)
+    }
+
+    /// Convert to a signed authorization by adding a signature.
+    pub const fn into_signed<S>(self, signature: S) -> SignedAuthorization<S> {
+        SignedAuthorization { inner: self, signature }
+    }
+}
+
+/// A signed EIP-7702 authorization.
+#[derive(Debug, Clone, RlpEncodable, RlpDecodable)]
+pub struct SignedAuthorization<S> {
```

Comment:
Can we add  `RecoveredAuthorization`? Where we will cache the recovered account as Option<Address> (Option as it can fail).
Recovery can be expensive and can be done before revm gets called. We are okay to do that as it is covered by gas.

Idea that I have is to have a enum as in Revm input:
```
AuthorisationList {
   Recovered(Vec<RecoveredAuthorization>),
   SignedAuthorization(Vec<SignedAuthorization>),
}
```

Reply:
added